[Feature] Institution Login for OKState University and Delegate Authentication to External CAS Server [#CAS-15][#OSF-7271]

cslzchen opened the pull request:
### Deploy Notes:

- update `cas.properties`
- update `institutions-auth.xsl`

OK State University:
```
<!-- OK State University -->
<xsl:when test="$idp='okstate'">
    <id>okstate</id>
    <user>
        <username><xsl:value-of select="//attribute[@name='mail']/@value"/></username>
        <fullname/>
        <familyName><xsl:value-of select="//attribute[@name='givenName']/@value"/></familyName>
        <givenName><xsl:value-of select="//attribute[@name='sn']/@value"/></givenName>
        <middleNames/>
        <suffix/>
    </user>
</xsl:when>
```

Duke University
```
<!-- Duke University -->
<xsl:when test="$idp='<replace_with_the_idp>'">
    <id>duke</id>
    <user>
        <username><xsl:value-of select="//attribute[@name='mail']/@value"/></username>
        <fullname><xsl:value-of select="//attribute[@name='displayName']/@value"/></fullname>
        <familyName><xsl:value-of select="//attribute[@name='givenName']/@value"/></familyName>
        <givenName><xsl:value-of select="//attribute[@name='sn']/@value"/></givenName>
        <middleNames/>
        <suffix/>
    </user>
</xsl:when>
```

### Tickets and PRs:

- Replaces #30, #37 
- OSF PR: https://github.com/CenterForOpenScience/osf.io/pull/7120
- CAS Ticket: https://openscience.atlassian.net/browse/CAS-15
- OSF Ticket: https://openscience.atlassian.net/browse/OSF-7271

cslzchen committed:
Update cas settings for OkState auth delegation.

cslzchen committed:
Update spring configs for OkState delegation login.

cslzchen committed:
Overwrite ClientAuthenticationHandler to add unique client name
 to the provider part of the id when creating a Principal:
 e.g. CasClientOkStateProfile#xxx-xx-xxxx

cslzchen committed:
Overwrite pac4j class ClientAction to handle institution
 and non-instituion clients:
 - institution clients go to remoteAuthenticate and
   authenticate through API
 - non-institution clients follow normal delegation flow

cslzchen committed:
Update pac4j class ClientAuthenticationHandler to
 customize principal id for institution login purpose.

cslzchen committed:
Update remoteAuthenticate to handle both SAML and CAS for
 institution login and update spring config.

cslzchen committed:
Update cas config, use stgcas and remove test fakecas.

cslzchen committed:
Add okstateu login url to casInstitutionLoginView.

cslzchen committed:
Use `okstate` as client name for OKStateUniversity:
 and client name must be identicial to institution id.

cslzchen committed:
Overlay org.pac4j.cas.client.CasClient.

cslzchen committed:
Add CAS 3.0 support to pac4j 1.7 and cas 4.1.

cslzchen committed:
Update CasClient with configurable protocol
 and fix style.

cslzchen committed:
Update version docs.

cslzchen committed:
Use okstateUrl instead of okstateuUrl given client name change.

cslzchen committed:
Handle released attributes.

cslzchen committed:
Merge remote-tracking branch 'longze/feature/institution-login-using-cas-delegation' into feature/instituion-login-okstate.

cslzchen committed:
Update institution model and daos for delegation protocol.

cslzchen committed:
Update institution handler (front and back end).

cslzchen committed:
Use static property name instead of literal String.

cslzchen committed:
Remove CAS30 support and unnecesarry overlay files,
    knowning that it is easy to add if we need it in the future.

cslzchen committed:
Fix institution daos and handler.

cslzchen committed:
Update docs on how and why we customize the building of  principal.

cslzchen committed:
Clean up code.

cslzchen committed:
Merge remote-tracking branch 'upstream/feature/django-osf' into feature/instituion-login-okstate

cslzchen committed:
Add cas.okstate.cas.protocol=SAML to settings.

cslzchen committed:
Merge remote-tracking branch 'upstream/develop' into feature/institution-login-okstate

cslzchen committed:
Merge remote-tracking branch 'upstream/develop'
 into feature/institution-login-okstate.

cslzchen committed:
Use XML XSLT.
    - update credential's authentication headers with
      attributes from the principal constructed from
      pac4j-cas authentication
    - replace manual parsing with XML XSLT

cslzchen committed:
Update comments and docstr.

cslzchen commented on cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java:
```diff
@@ -269,65 +279,102 @@ protected Credential constructCredentialsFromRequest(final RequestContext contex
         // rejects any conflicting / forged headers.
         final String shibbolethSession = request.getHeader(SHIBBOLETH_SESSION_HEADER);
+
         if (StringUtils.hasText(shibbolethSession)) {
```
Institution Login by External Shibboleth IDP.

cslzchen commented on cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java:
```diff
@@ -269,65 +279,102 @@ protected Credential constructCredentialsFromRequest(final RequestContext contex
             credential.setInstitutionId(remoteUserInfo.getInstitutionId());
-
+            return credential;
+        } else if (context.getFlowScope().get("authenticationDelegationProtocol") == PROTOCOL_CAS) {
```
Institution Login by External CAS IDP.

cslzchen commented on cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java:
```diff
@@ -269,65 +279,102 @@ protected Credential constructCredentialsFromRequest(final RequestContext contex
+            }
+            try {
+                credential.setRemotePrincipal(Boolean.TRUE);
+                principal = ticketGrantingTicket.getAuthentication().getPrincipal();
```
Retrieve the `principal` from `TGT` which contains authenticated institution user's information.

cslzchen commented on cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java:
```diff
@@ -409,13 +455,45 @@ private JSONObject normalizeRemotePrincipal(final OpenScienceFrameworkCredential
+     * @return the updated credential
+     * @throws AccountException RemoteUserFailedLoginException
+     */
+    private OpenScienceFrameworkCredential constructCredentialFromRemotePrincipal(
```
Update `credential.authenticationHeaders` which will be normalized with XML XSLT.
- `Shib-Identity-Provider`
- `uid`
- attributes in `principal`
  - OK State University
    - `mail` is mapped to `username` (they don't provide `eppn`)
    - `fullname` is mapped to `uid` (they don't provider `displayName`)
    - `sn` is mapped to `familyName`
    - `givenName` is mapped to `givenName`

cslzchen commented on cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java:
```diff
@@ -269,65 +279,102 @@ protected Credential constructCredentialsFromRequest(final RequestContext contex
+            if (PROTOCOL_SAML.equals(protocol)) {
+                normalizedPayload = this.normalizeRemotePrincipal(credential);
+            } else if (PROTOCOL_CAS.equals(protocol)) {
+                normalizedPayload = this.normalizeRemotePrincipal((this.constructCredentialFromRemotePrincipal(credential, principal)));
```
For CAS IDP, `credential` does not contain authenticated user's information yet. Call `constructCredentialFromRemotePrincipal` to update it with `principal` before normalization.

cslzchen commented on cas-server-webapp/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/ClientAuthenticationHandler.java:
```diff
@@ -0,0 +1,111 @@
+     *      institution login flow after successful auth. The principal id only contains client name and profile id.
+     */
+    @Override
+    protected HandlerResult createResult(final ClientCredential credentials, final UserProfile profile)
```
The default method creates a principal after successful authentication. The principal is created with the id that uses class name (e.g. `CasClient`), in which case we can not identify the client name (e.g. `okstate`). We override it to use the client name.

cslzchen commented on cas-server-webapp/src/main/java/org/jasig/cas/support/pac4j/web/flow/ClientAction.java:
```diff
@@ -0,0 +1,258 @@
+                WebUtils.putTicketGrantingTicketInScopes(context, tgt);
+
+                // for institution clients, go to `remoteAuthenticate` and finish authentication through API
+                if (INSTITUTION_CLIENTS.contains(clientName)) {
```
After successfully authentication, all necessary information is already stored in the `principal` in the `TGT`. If we identify the client to be an institution client, the flow is intercepted and redirected to our institution login flow independent of the delegation protocol.

cslzchen commented on cas-server-webapp/src/main/webapp/WEB-INF/webflow/login/login-webflow.xml:
```diff
@@ -34,6 +34,7 @@
         <evaluate expression="clientAction" />
         <transition on="success" to="sendTicketGrantingTicket" />
         <transition on="error" to="ticketGrantingTicketCheck" />
+        <transition on="remote" to="remoteAuthenticate" />
```
As mentioned before, the spring config for intercepted flow for institution login.

cslzchen commented on cas-server-webapp/src/main/webapp/WEB-INF/view/jsp/default/ui/casInstitutionLoginView.jsp:
```diff
@@ -50,6 +50,8 @@
             if(login_url == null || login_url == "") {
                 return;
+            } else if (login_url == "okstate") {
+                login_url = "${okstateUrl}";
```
As mentioned before, dynamically build `login_url`.

cslzchen commented on cas-server-support-osf/src/main/java/io/cos/cas/adaptors/postgres/handlers/OpenScienceFrameworkInstitutionHandler.java:
```diff
@@ -64,17 +64,41 @@ public String findInstitutionLogoutUrlById(final String id) {
+                if ("saml-shib".equals(institution.getDelegationProtocol())) {
+                    institutionLogin.put(institution.getLoginUrl() + "&target=" + target, institution.getName());
+                } else if ("cas-pac4j".equals(institution.getDelegationProtocol())) {
+                    institutionLogin.put(institution.getId(), institution.getName());
```
Login URLs using `pac4j` are built dynamically. This is why we set login_url to be the client_name for OK State University.

cslzchen committed:
Merge remote-tracking branch 'upstream/develop'
 into feature/institution-login-okstate.

cslzchen commented on cas-server-support-osf/src/main/java/io/cos/cas/adaptors/postgres/daos/OpenScienceFrameworkDaoImpl.java:
```diff
@@ -127,7 +127,8 @@ public OpenScienceFrameworkInstitution findOneInstitutionById(final String id) {
             final TypedQuery query = entityManager.createQuery(
-                    "select i from OpenScienceFrameworkInstitution i where i.loginUrl is not null and i.deleted = false",
+                    "select i from OpenScienceFrameworkInstitution i "
+                            + "where i.delegationProtocol is not empty and i.deleted = false",
```
`delegation_protocol` is used to determine whether an institution has login available, instead of `login_url`.

icereval commented on cas-server-webapp/src/main/java/org/jasig/cas/support/pac4j/web/flow/ClientAction.java:
```diff
@@ -0,0 +1,258 @@
+public final class ClientAction extends AbstractAction {
+
+    /** The set INSTITUTION_CLIENTS contains all clients whose authentication is considered as Institution Login. */
+    public static final Set<String> INSTITUTION_CLIENTS = ImmutableSet.of(
```
institutions should not be hardcoded, should perform a database lookup

cslzchen committed:
Use enum instead of literal string for delegation protocols.

cslzchen committed:
Add `institutionHandler` to `ClientAction` and `ClientAuthenticationHandler`
 for them to access institution model and check if a client is an institution.

cslzchen committed:
Update model for `osf_apioauth2application`.

cslzchen commented on cas-server-support-osf/src/main/java/io/cos/cas/adaptors/postgres/models/OpenScienceFrameworkInstitution.java:
```diff
@@ -34,6 +34,14 @@
 public class OpenScienceFrameworkInstitution {
 
+    /** A set of supported institution delegation protocol with implementation information. */
+    public enum DelegationProtocols {
```
Use `enum` instead of hard-coded literal `String`.

cslzchen commented on cas-server-webapp/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/ClientAuthenticationHandler.java:
```diff
@@ -0,0 +1,121 @@
+
+    /** The handler for osf institution. */
+    @NotNull
+    private final OpenScienceFrameworkInstitutionHandler institutionHandler;
```
Need an instance of `OpenScienceFrameworkInstitutionHandler` to access ` OpenScienceFrameworkInstitution` model and dao.

cslzchen commented on cas-server-webapp/src/main/java/org/jasig/cas/support/pac4j/web/flow/ClientAction.java:
```diff
@@ -0,0 +1,272 @@
+
+    /** The handler for osf institution. */
+    @NotNull
+    private final OpenScienceFrameworkInstitutionHandler institutionHandler;
```
Need an instance of `OpenScienceFrameworkInstitutionHandler` to access ` OpenScienceFrameworkInstitution` model and dao.

cslzchen commented on cas-server-webapp/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/ClientAuthenticationHandler.java:
```diff
@@ -0,0 +1,121 @@
+
+        if (clientName != null) {
+            // customize profile names
+            if (this.institutionHandler.isDelegatedInstitutionLogin(clientName)) {
```
Identity clients that are also institution.

cslzchen commented on cas-server-webapp/src/main/java/org/jasig/cas/support/pac4j/web/flow/ClientAction.java:
```diff
@@ -0,0 +1,272 @@
+                WebUtils.putTicketGrantingTicketInScopes(context, tgt);
+
+                // for institution clients, go to `remoteAuthenticate` and finish authentication through API
+                if (this.institutionHandler.isDelegatedInstitutionLogin(clientName)) {
```
Identify clients that are also institution. Intercept the flow and go to `remoteAuthenticate`.

cslzchen commented on cas-server-support-osf/src/main/java/io/cos/cas/adaptors/postgres/models/OpenScienceFrameworkApiOauth2Application.java:
```diff
@@ -44,7 +44,7 @@
      * Methods such as `getId()` refer to this `objectId`.
      */
-    @Column(name ="_id", nullable = false)
+    @Column(name ="_id", nullable = false, unique=true)
```
Stay consistent with the OSF database.

cslzchen commented:
After we enforce the `fullname` requirement, the test account for okstate can no longer sign in. If you see the following log/sentry message for API server, consider authentication successful.
```
[framework.sentry]  WARNING: Sentry called to log message, but is not active: Institution login failed: fullname required for user test.user14@okstate.edu from institution okstate
```

cslzchen committed:
Add mock class for CasClient, OrcidClient and OSFInstitutionHandler.

cslzchen committed:
Add unit tests for ClientAuthenticationHandler.

cslzchen committed:
Add JUnit dependency for cas-server-webapp.

icereval commented on cas-server-webapp/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/ClientAuthenticationHandler.java:
```diff
@@ -0,0 +1,121 @@
+     *      For OSF compatibility, only CAS clients use client name while the ORCiD client is not changed
+     * 2.   For clients considered as "institution", their auth flow is intercepted and redirected to our
+     *      institution login flow after successful auth. The principal id only contains client name and profile id.
+     */
```
please add examples of the `clientName` cases

icereval commented on cas-server-webapp/src/test/java/org.jasig.cas.support.pac4j/authentication.handler.support/ClientAuthenticationHandlerTests.java:
```diff
@@ -0,0 +1,151 @@
+/*
```
Add test to validate two factor auth w/ each provider.

icereval commented on cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java:
```diff
@@ -409,13 +455,45 @@ private JSONObject normalizeRemotePrincipal(final OpenScienceFrameworkCredential
+
+        // use institution id for CAS idp
+        final String institutionId = principal.getId().split("#")[0];
+        credential.getAuthenticationHeaders().put("Shib-Identity-Provider", institutionId);
```
Rename `getAuthenticationHeders` to a more general term like `getDelegationAttributes`.

Change to `Shib-Identity-Provider` to `Cas-Identity-Provider`.

Add `Delegation-Protocol` to `credential` for both CAS and SAML and append to normalization document xslt processor.

icereval commented on cas-server-webapp/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/ClientAuthenticationHandler.java:
```diff
@@ -0,0 +1,121 @@
+                // institution clients are independent of authentication delegation protocol
+                // principal id contains client name and profile id
+                id = clientName + '#' + profile.getId();
+            } else if (clientName.startsWith("CasClient")) {
```
remove this case for now, we can dig into what we would like this to be named in the future

icereval commented on cas-server-webapp/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/ClientAuthenticationHandler.java:
```diff
@@ -0,0 +1,121 @@
+            if (this.institutionHandler.isDelegatedInstitutionLogin(clientName)) {
+                // institution clients are independent of authentication delegation protocol
+                // principal id contains client name and profile id
+                id = clientName + '#' + profile.getId();
```
use standard naming convention w/ `Profile` designation at the end.

e.g. `id = clientName + "Profile#" + profile.getId();`

icereval commented on cas-server-support-osf/src/main/java/io/cos/cas/adaptors/postgres/handlers/OpenScienceFrameworkInstitutionHandler.java:
```diff
@@ -60,16 +64,44 @@ public String findInstitutionLogoutUrlById(final String id) {
         for (final OpenScienceFrameworkInstitution institution: institutionList) {
-            institutionLogin.put(institution.getLoginUrl() + "&target=" + target, institution.getName());
+            if (institution.getDelegationProtocol().equalsIgnoreCase(
+                    OpenScienceFrameworkInstitution.DelegationProtocols.SAML_SHIB.name())) {
```
in a future pr we should consider moving shibboleth url construction logic to the cas side

icereval commented on cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java:
```diff
@@ -409,13 +455,45 @@ private JSONObject normalizeRemotePrincipal(final OpenScienceFrameworkCredential
+        final String institutionId = principal.getId().split("#")[0];
+        credential.getAuthenticationHeaders().put("Shib-Identity-Provider", institutionId);
+
+        // for institutions that do not release full name, set uid as their full name in "institution-auth.xsl"
```
???

cslzchen committed:
Add ClientActionTests to verify dynamically
generated login url for okstate and orcid.

cslzchen committed:
Update pom and docstr.

cslzchen committed:
Rename authenticationHeaders to delegationAttributes in
OpenScienceFrameworkCredentials.

cslzchen committed:
Institution by CAS_PAC4J uses only clientName as princial id,
remove unsupported non-institution by CAS_PAC$J, and add
Delegation-Protocol to delegationAttributes for xslt parsing.

cslzchen committed:
Update tests.

cslzchen committed:
Rename pac4j beans.

cslzchen committed:
Update docstr for pac4j customization. [skip ci]

cslzchen committed:
Use OSFInstitution.DelegationProtocols enum and
fix styel [skip ci]

cslzchen committed:
Update credential with institution and delegation fields
during cas-pac4j authentication, add delegationProtocol
to OpenScienceFrameworkCredential, and simplify
notifyRemotePrincipalAuthenticated().

cslzchen commented on cas-server-support-osf/src/main/java/io/cos/cas/authentication/OpenScienceFrameworkCredential.java:
```diff
@@ -62,8 +63,11 @@
-    /** The Authentication Headers. */
-    private Map<String, String> authenticationHeaders = new HashMap<>();
+    /** The Authentication Delegation Protocol. */
+    private String delegationProtocol;
```
Add `delegationProtocol` (above) and rename `authenticationHeaders` to `delegationAttributes` (below) for `OpenScienceFrameworkCredential`, where now there are four institution related properties:
- `remotePrincipal`
- `institutionId`
- `delegationProtocol`
- `delegationAttributes`

cslzchen commented on cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java:
```diff
@@ -113,21 +128,21 @@ public PrincipalAuthenticationResult(final String username, final String institu
     }
 
+    /** Authentication Delegation Protocol: CAS by pac4j. */
+    public static final String PROTOCOL_CAS = OpenScienceFrameworkInstitution.DelegationProtocols.CAS_PAC4J.name();
```
Uses `enum` instead of literal `String` for `PROTOCOL_CAS` (above) and same for `PROTOCOL_SAML` (below).

cslzchen commented on cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java:
```diff
@@ -269,65 +281,102 @@ protected Credential constructCredentialsFromRequest(final RequestContext contex
                 }
             }
-
+            credential.getDelegationAttributes().put("Delegation-Protocol", PROTOCOL_SAML);
```
(Above) Add `Delegation-Protocol` to `delegationAttributes` to improve `xslt`.
(Below) Update `credential` with the newly introduced `delegationProtocol`.

cslzchen commented on cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java:
```diff
@@ -269,65 +281,102 @@ protected Credential constructCredentialsFromRequest(final RequestContext contex
             credential.setInstitutionId(remoteUserInfo.getInstitutionId());
-
+            return credential;
+        } else if (PROTOCOL_CAS.equals(credential.getDelegationProtocol())) {
```
From flow action `clientAction` to `remoteAuthenticate` during `CAS_PAC4J` institution login, the `credential` contains the protocol information and institution id.

cslzchen commented on cas-server-support-osf/src/main/java/io/cos/cas/authentication/handler/support/OpenScienceFrameworkPrincipalFromRequestRemoteUserNonInteractiveCredentialsAction.java:
```diff
@@ -393,29 +442,57 @@ private JSONObject normalizeRemotePrincipal(final OpenScienceFrameworkCredential
+     * @return the updated credential
+     * @throws AccountException RemoteUserFailedLoginException
+     */
+    private OpenScienceFrameworkCredential constructCredentialFromRemotePrincipal(
```
Update `credential.delegationAttributes`:
- `Delegation-Protocol`: a control variable for XSLT to choose which type of transformation to apply.
- `Cas-Identity-Provider`: use `clientName` which is identical to `institutionId`
- `attributes` in `principal`
  - OK State University
    - `mail` is mapped to `username`
    - `sn` is mapped to `familyName`
    - `givenName` is mapped to `givenName`
- Refer to the `institutions-auth.xsl` I airdropped to you.

cslzchen commented on cas-server-webapp/src/main/java/org/jasig/cas/support/pac4j/web/flow/ClientAction.java:
```diff
@@ -0,0 +1,285 @@
+                        this.centralAuthenticationService.createTicketGrantingTicket(new ClientCredential(credentials));
+                WebUtils.putTicketGrantingTicketInScopes(context, tgt);
+
+                // for institution clients:
```
As commented below.

cslzchen commented on cas-server-webapp/src/main/java/org/jasig/cas/support/pac4j/authentication/handler/support/ClientAuthenticationHandler.java:
```diff
@@ -0,0 +1,134 @@
+    /**
+     * {@inheritDoc}
+     *
+     * Note on customization:
```
The updated docstr with examples.

cslzchen reviewed (commented):
@icereval ready with several updates

cslzchen committed:
Minor style fixes.

cslzchen commented on cas-server-support-osf/src/main/java/io/cos/cas/adaptors/postgres/handlers/OpenScienceFrameworkInstitutionHandler.java:
```diff
@@ -60,16 +60,45 @@ public String findInstitutionLogoutUrlById(final String id) {
+     */
+    public boolean isDelegatedInstitutionLogin(final String clientName) {
+        final OpenScienceFrameworkInstitution institution
+                = openScienceFrameworkDao.findOneInstitutionById(clientName.toLowerCase());
```
I think I misunderstood this before. Here is the `toLowerCase()` comparison for `clientName` as `institutionId`.

icereval committed:
collapse pac4j login into osf remote authentication handler

cslzchen committed:
Merge pull request #2 from icereval/feature/institution-login-okstate

collapse pac4j login into osf remote authentication handler

icereval committed:
handle delegation protocol enum error

cslzchen committed:
Merge pull request #3 from icereval/feature/institution-login-okstate

handle delegation protocol exception

cslzchen committed:
Fix style.

icereval commented:
Testing

1. master
2. new cas db
3. login w/ normal methods + oauth application
4. upgrade to pr
5. restart cas
6. verify methods remain logged in and you can access `/login` and api oauth endpoints w/ `Bearer <token>`

icereval committed:
Update OpenScienceFrameworkCredential.java

cslzchen committed:
Fix `null` comparison: credential.getDelegationProtocol().

cslzchen commented:
@icereval local tests done, all passed

### `develop` (identical to `master` for CAS)

* New CAS Database
* Four Users:
  * NYU user, with 2FA on, logged in
  * ORCID user, with 2FA on, logged in
  * OSF user, with 2FA on, logged in
  * OSF master, with developer app, with a valid PAT, logged in
* OSF user authorizes the Master App and Master App receives the AT, can profile and act on behalf of OSF user

### `feature/okstate` upgrade

* Perform Upgrade
* CAS starts successfully
* All checks below are verified:
  * All four users' CAS page still logged in. (TGT is still valid)
  * All four users' OSF page still logged in.
  * Log all four users out and login again successfully with 2FA working.
  * Master user's PAT still works. (PAT is still valid)
  * Master user's developer app gets loaded in CAS.
  * Master App can still profile and and act on behalf of the user. (AT is still valid)
  * NYU and ORCID users successfully authorize Master App.
* OKState
  * User successfully logged in.
  * 2FA works as expected.
